consoles/video_stream.pm

```diff
@@ -0,0 +1,350 @@
+        # typical max_interval values are
+        #   4ish:  veeery slow
+        #   15ish: slow
+        $seconds_per_keypress = $seconds_per_keypress + 1 / sqrt($args->{max_interval});
+    }
+
+    for my $letter (split("", $args->{text})) {
+        next if ($letter eq "\r");
+        $letter = $CHARMAP->{$letter} || $letter;
+        $self->_send_key_event($letter);
+        $self->{backend}->run_capture_loop($seconds_per_keypress);
+    }
+    return {};
+}
+
+sub send_key ($self, $args) {
+    $self->_send_key_event($args->{key});
+    $self->backend->run_capture_loop(.2);
+    return {};
+}
+
+sub _send_key_event ($self, $key) {
+    return unless $self->{input_pipe};
+    $self->{input_pipe}->write($key . "\n")
+      or die "failed to send '$key' input event";
+}
+
+sub get_last_mouse_set ($self, @) {
+    return $self->{mouse};
+}
+
+sub _mouse_move ($self, $x, $y) {
+    die "need parameter \$x and \$y" unless (defined $x and defined $y);
+    return unless $self->{input_pipe};
+
+    if ($self->{mouse}->{x} == $x && $self->{mouse}->{y} == $y) {
+        # in case the mouse is moved twice to the same position
+        # (e.g. in case of duplicated mouse_hide), we need to wiggle the
+        # mouse a bit to avoid qemu ignoring the repositioning
+        # because the SUT might have moved the mouse itself and we
+        # need to make sure the mouse is really where expected
+        my $delta = 5;
+        # move it to the left in case the mouse is right
+        #$delta = -5 if $x > $self->{width} / 2;
+        $delta = -5 if $x > 1024 / 2;
+        $self->_mouse_move($x + $delta, $y);
+    }
+
+    bmwqemu::diag "mouse_move to $x, $y";
+    $self->{input_pipe}->write("mouse_move $x $y\n");
+    $self->{input_pipe}->flush;
+
+    $self->{mouse}->{x} = $x;
+    $self->{mouse}->{y} = $y;
+    # let the event be processed before further commands
+    $self->backend->run_capture_loop(.1);
+
+    return;
+}
+
+sub mouse_hide ($self, $args) {
+    $args->{border_offset} //= 0;
+
+    #my $x = $self->{width} - 1;
+    #my $y = $self->{height} - 1;
+    my $x = 1024 - 1;
+    my $y = 768 - 1;
+
+    if (defined $args->{border_offset}) {
+        my $border_offset = int($args->{border_offset});
+        $x -= $border_offset;
+        $y -= $border_offset;
+    }
+
+    $self->_mouse_move($x, $y);
+    return {absolute => $self->{mouse}};
+}
+
+sub mouse_set ($self, $args) {
+    die "Need x/y arguments" unless (defined $args->{x} && defined $args->{y});
+
+    $self->_mouse_move(int($args->{x}), int($args->{y}));
+    return {};
+}
+
+sub mouse_button ($self, $args) {
+    return unless $self->{input_pipe};
+    my $button = $args->{button};
+    my $bstate = $args->{bstate};
+
+    my $mask = 0;
+    if ($button eq 'left') {
+        $mask = $bstate;
+    }
+    elsif ($button eq 'right') {
+        $mask = $bstate << 2;
+    }
+    elsif ($button eq 'middle') {
+        $mask = $bstate << 1;
+    }
```

Comment:
It would be nice *not* to C&P this code from the VNC console (especially since the code will be improved by another pending PR).

Reply:
Do you suggest a different API for sending button presses, or just different way to write this part?

Reply:
I suggest to avoid code duplication because this code looks very familiar.

Reply:
Technically, it is a different protocol, even if indeed both have mouse buttons as a bitmask. For example VNC uses a single command for both mouse move and click, while this one have separate commands. But I can make the improvement similar to VNC's here too.

Reply:
Sure, this is not about the difference in protocol but just simply the technical implementation regarding OOP. Please see https://github.com/os-autoinst/os-autoinst/pull/1898#discussion_r786487750